source/f5_transportserver.go

```diff
@@ -0,0 +1,213 @@
+
+	// wait for the local cache to be populated.
+	if err := waitForDynamicCacheSync(context.Background(), informerFactory); err != nil {
+		return nil, err
+	}
+
+	uc, err := newTSUnstructuredConverter()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to setup unstructured converter")
+	}
+
+	return &f5TransportServerSource{
+		dynamicKubeClient:       dynamicKubeClient,
+		transportServerInformer: transportServerInformer,
+		kubeClient:              kubeClient,
+		namespace:               namespace,
+		annotationFilter:        annotationFilter,
+		unstructuredConverter:   uc,
+	}, nil
+}
+
+// Endpoints returns endpoint objects for each host-target combination that should be processed.
+// Retrieves all TransportServers in the source's namespace(s).
+func (ts *f5TransportServerSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	transportServerObjects, err := ts.transportServerInformer.Lister().ByNamespace(ts.namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	var transportServers []*f5.TransportServer
+	for _, tsObj := range transportServerObjects {
+		unstructuredHost, ok := tsObj.(*unstructured.Unstructured)
+		if !ok {
+			return nil, errors.New("could not convert")
+		}
+
+		transportServer := &f5.TransportServer{}
+		err := ts.unstructuredConverter.scheme.Convert(unstructuredHost, transportServer, nil)
+		if err != nil {
+			return nil, err
+		}
+		transportServers = append(transportServers, transportServer)
+	}
+
+	transportServers, err = ts.filterByAnnotations(transportServers)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to filter TransportServers")
+	}
+
+	endpoints, err := ts.endpointsFromTransportServers(transportServers)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort endpoints
+	for _, ep := range endpoints {
+		sort.Sort(ep.Targets)
+	}
+
+	return endpoints, nil
+}
+
+func (ts *f5TransportServerSource) AddEventHandler(ctx context.Context, handler func()) {
+	log.Debug("Adding event handler for TransportServer")
+
+	ts.transportServerInformer.Informer().AddEventHandler(eventHandlerFunc(handler))
+}
+
+// endpointsFromTransportServers extracts the endpoints from a slice of TransportServers
+func (ts *f5TransportServerSource) endpointsFromTransportServers(transportServers []*f5.TransportServer) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+
+	for _, transportServer := range transportServers {
+		resource := fmt.Sprintf("f5-transportserver/%s/%s", transportServer.Namespace, transportServer.Name)
+
+		ttl := getTTLFromAnnotations(transportServer.Annotations, resource)
+
+		targets := getTargetsFromTargetAnnotation(transportServer.Annotations)
+		if len(targets) == 0 && transportServer.Spec.VirtualServerAddress != "" {
+			targets = append(targets, transportServer.Spec.VirtualServerAddress)
+		}
+		if len(targets) == 0 && transportServer.Status.VSAddress != "" {
+			targets = append(targets, transportServer.Status.VSAddress)
+		}
```

Comment:
Wdyt of detailing this behavior in the tutorial ?

Reply:
LMK if the examples I added are enough.